### ISSUE TITLE
iotedged 1.1.6 fails to remove & start edgeAgent if edgeAgent container is in "Dead" state

### ISSUE THREAD

**Issue by derSchtefan** [Azure/iotedge#5514]
#### TL;DR

iotedged 1.1.6 does not handle edgeAgent container state `Dead` correctly and instead of trying a removal and recreation, insists on starting it, fails in a loop. Needs to check edgeAgent container state before applying "start" command.

#### Details

As part of scheduled maintenance,  `apt upgrade` performed an upgrade of the iotedged and the moby container runtime. During this both services restarted. Afterwards the system was rebooted for a kernel update.

When iotedged restarts, it removes all docker containers, creates the edgeAgent specified in the yaml file (1.1, in this case, the default), and then lets that edge agent recreate all docker containers.

There seems to be a flaw in the logic how iotedged determines which docker action to perform: It appears that if the edgeAgent is in state "Dead", the container is neither in stopped nor running. Iotedged will thus not remove the container but try to start the Dead container over and over.

The edgeAgent container seems to have slipped into the `Dead` state because moby was restarting while iotedged was trying to remove edgeAgent 1.1 as part of spinning up edgeAgent:1.1.4. You see `connection closed`, `Error 500`, and `Broken pipe` around that time. This is the log in question:

```
iotedged[13891]: 2021-09-15T11:42:53Z [INFO] - Pulling image mcr.microsoft.com/azureiotedge-agent:1.1.4...
iotedged[13891]: 2021-09-15T11:42:54Z [INFO] - Successfully pulled image mcr.microsoft.com/azureiotedge-agent:1.1.4
iotedged[13891]: 2021-09-15T11:42:54Z [INFO] - [mgmt] - - - [2021-09-15 11:42:54.051184399 UTC] "POST /modules/edgeAgent/prepareupdate?api-version=2020-07-07 HTTP/1.1" 204 No Content - "-" "-" auth_i
iotedged[13891]: 2021-09-15T11:42:54Z [INFO] - Updating and starting module edgeAgent
iotedged[13891]: 2021-09-15T11:42:54Z [INFO] - Removing module edgeAgent...
iotedged[13891]: 2021-09-15T11:42:55Z [WARN] - Could not remove module edgeAgent
iotedged[13891]: 2021-09-15T11:42:55Z [WARN] -         caused by: Container runtime error
iotedged[13891]: 2021-09-15T11:42:55Z [WARN] -         caused by: connection closed before message completed
iotedged[13891]: 2021-09-15T11:42:55Z [ERR!] - Internal server error: Could not update module "edgeAgent"
iotedged[13891]:         caused by: Could not remove module edgeAgent
iotedged[13891]:         caused by: Container runtime error
iotedged[13891]:         caused by: connection closed before message completed
iotedged[13891]: 2021-09-15T11:42:55Z [INFO] - [mgmt] - - - [2021-09-15 11:42:55.094946607 UTC] "PUT /modules/edgeAgent?api-version=2020-07-07&start=true HTTP/1.1" 500 Internal Server Error 194 "-" "
iotedged[13891]: 2021-09-15T11:42:55Z [ERR!] - server connection error: (unknown)
iotedged[13891]: 2021-09-15T11:42:55Z [ERR!] - error writing a body to connection: Broken pipe (os error 32)
iotedged[13891]: 2021-09-15T11:43:48Z [INFO] - Checking edge runtime status
iotedged[13891]: 2021-09-15T11:43:48Z [INFO] - Edge runtime status is failed, starting module now...
iotedged[13891]: 2021-09-15T11:43:48Z [INFO] - Starting module edgeAgent...
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] - Could not start module edgeAgent
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] -         caused by: Conflict with current operation
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] - Error in watchdog when checking for edge runtime status:
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] - A module runtime error occurred.
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] -         caused by: Could not start module edgeAgent
iotedged[13891]: 2021-09-15T11:43:48Z [WARN] -         caused by: Conflict with current operation
iotedged[13891]: 2021-09-15T11:43:48Z [INFO] - Removing listener for module edgeAgent
iotedged[13891]: 2021-09-15T11:43:48Z [INFO] - Stopping listener for module edgeAgent
iotedged[13891]: 2021-09-15T11:44:14Z [INFO] - Received SIGTERM, starting shutdown
iotedged[13891]: 2021-09-15T11:44:14Z [INFO] - Stopping edge runtime module edgeAgent
iotedged[13891]: 2021-09-15T11:44:14Z [INFO] - Stopping module edgeAgent...
iotedged[13891]: 2021-09-15T11:44:14Z [WARN] - Could not stop module edgeAgent
iotedged[13891]: 2021-09-15T11:44:14Z [WARN] -         caused by: Target of operation already in this state
iotedged[13891]: 2021-09-15T11:44:14Z [ERR!] - Edge runtime will stop because the shutdown signal raised an error.
systemd[1]: Stopping Azure IoT Edge daemon...
iotedged[13891]: 2021-09-15T11:44:14Z [ERR!] - The watchdog encountered an error
iotedged[13891]: 2021-09-15T11:44:14Z [ERR!] -         caused by: A module runtime error occurred.
iotedged[13891]: 2021-09-15T11:44:14Z [ERR!] -         caused by: Could not stop module edgeAgent
iotedged[13891]: 2021-09-15T11:44:14Z [ERR!] -         caused by: Target of operation already in this state
iotedged[13891]: 2021-09-15T11:42:47Z [INFO] (/project/edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_log.c:log_init:41) Initialized logging
iotedged[13891]: 2021-09-15T11:42:47Z [INFO] (/project/edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_crypto.c:edge_hsm_client_destroy_certificate:344) Could not destroy certificate in the store
iotedged[13891]: 2021-09-15T11:42:47Z [INFO] (/project/edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_store.c:edge_hsm_client_store_create_pki_cert:2706) Generating certificate and key for alias
iotedged[13891]: 2021-09-15T11:42:47Z [INFO] (/project/edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:generate_rsa_key:346) Generating RSA key of length 4096
systemd[1]: iotedge.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: iotedge.service: Failed with result 'exit-code'.
systemd[1]: Stopped Azure IoT Edge daemon.
-- Reboot --
systemd[1]: Started Azure IoT Edge daemon.
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Starting Azure IoT Edge Security Daemon
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Version - 1.1.6
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Using config file: /etc/iotedge/config.yaml
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Configuring /var/lib/iotedge as the home directory.
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Configuring certificates...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Transparent gateway certificates not found, operating in quick start mode...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Finished configuring provisioning environment variables and certificates.
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Initializing hsm...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Finished initializing hsm.
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Provisioning edge device...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Starting provisioning edge device via manual mode using a device connection string...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Manually provisioning device "xxx" in hub "xxxxxx.azure-devices.net"
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Finished provisioning edge device.
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Initializing the module runtime...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Initializing module runtime...
iotedged[1796]: 2021-09-15T11:44:40Z [INFO] - Using runtime network id azure-iot-edge
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Successfully initialized module runtime
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Finished initializing the module runtime.
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Stopping all modules...
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Stopping module edgeAgent...
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] - Could not stop module edgeAgent
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] -         caused by: Target of operation already in this state
```

and later in a loop:
```
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Checking edge runtime status
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Edge runtime status is failed, starting module now...
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Starting module edgeAgent...
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] - Could not start module edgeAgent
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] -         caused by: Conflict with current operation
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] - Error in watchdog when checking for edge runtime status:
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] - A module runtime error occurred.
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] -         caused by: Could not start module edgeAgent
iotedged[1796]: 2021-09-15T11:44:41Z [WARN] -         caused by: Conflict with current operation
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Removing listener for module edgeAgent
iotedged[1796]: 2021-09-15T11:44:41Z [INFO] - Stopping listener for module edgeAgent
iotedged[1796]: 2021-09-15T11:45:41Z [INFO] - Checking edge runtime status
iotedged[1796]: 2021-09-15T11:45:41Z [INFO] - Edge runtime status is failed, starting module now...
iotedged[1796]: 2021-09-15T11:45:41Z [INFO] - Starting module edgeAgent...
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] - Could not start module edgeAgent
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] -         caused by: Conflict with current operation
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] - Error in watchdog when checking for edge runtime status:
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] - A module runtime error occurred.
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] -         caused by: Could not start module edgeAgent
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] -         caused by: Conflict with current operation
iotedged[1796]: 2021-09-15T11:45:41Z [INFO] - Removing listener for module edgeAgent
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] - Could not remove socket with uri unix:///var/lib/iotedge/mnt/edgeAgent.sock
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] - The workload manager encountered an error
iotedged[1796]: 2021-09-15T11:45:41Z [WARN] -         caused by: No such file or directory (os error 2)
iotedged[1796]: 2021-09-15T11:46:41Z [INFO] - Checking edge runtime status
```

If we have a look at the docker containers present:
```
docker ps -a
CONTAINER ID   IMAGE                                      COMMAND                  CREATED        STATUS    PORTS     NAMES
b5512681c366   mcr.microsoft.com/azureiotedge-agent:1.1   "/bin/sh -c 'exec /a…"   20 hours ago   Dead                edgeAgent
```

After we manually removed the dead container with `docker rm edgeAgent` iotedged brought back the whole system immediately. 

## Expected Behavior
The container state "Dead" of the edgeAgent needs to be handled. This state is documented as being potentially possible when a container has an operation pending that did not complete during a Docker/moby restart. That iotedged and moby get updated at the same time, and both restart, is not unlikely, given the fact that edge device updates are batched and executed in set maintenance windows - it is thus likely that both get updated and restart at some point together.

iotedged should query the state of the edgeAgent container, and handle states like "Dead" by removing and recreating the container.

## Steps to Reproduce
It is tricky, because it does not always happen. You need to manage to get the edgeAgent into the Dead state.
1. Deploy an IOT solution with a couple of larger docker containers that perform some action / delay when requested to shutdown 
1. Downgrade iotedged to a version lower than 1.1.6
2. Downgrade moby 
3. Run apt update && apt full-upgrade -y

If this is not enough you might want to do it again but reboot the system (this sometimes hits the right timing conditions)

## Workaround

The workarounds require manual intervention on the edge devices:
- Reboot / Power cycle the system (moby/Docker removes Dead containers upon reboot)
- SSH onto the device and `docker rm edgeAgent`

### Device Information
* Ubuntu 18.04

### Runtime Versions
* aziot-edged: 1.1.6
* Edge Agent: 1.1 in /etc/iotedge/config.yaml,  1.1.4 in deployed solution 
* Edge Hub 1.1.4
* Docker/Moby: engine: 20.10.8+azure, containerd: 1.4.9+azure, runc: 1.0.1, docker-init:  0.19.0

**Comment by gri6507** [comment]
After spending about a day troubleshooting this very issue on my unit, I have a keen interest in seeing this resolved. Do you have an ETA for the fix?

**Comment by gordonwang0** [comment]
I have merged the PR that resolved this issue. The fix will go out in the next release of iotedge.

**Comment by damonbarry** [comment]
> I have merged the PR that resolved this issue. The fix will go out in the next release of iotedge.

@derSchtefan @gri6507 To clarify, this fix has been merged to master. At this point, it will not go into the 1.1 LTS release. Rather, it will go into a future release of 1.2.

**Comment by derSchtefan** [comment]
@damonbarry We are about to go live with a world-wide IoT-edge solution based on 1.1 LTS *BECAUSE* it is a LTS. 1.2 has an incompatible file layout and config file format (config files which our software writes during device manufacturing).

Now you tell us that such a critical bug won't be fixed in something called a "LTS" version?

**Comment by gri6507** [comment]
> Now you tell us that such a critical bug won't be fixed in something called a "LTS" version?

@damonbarry - I have to echo @derSchtefan comment. We are also about to go live using 1.1.x because it is "LTS" and is incompatible with 1.2.x releases. With this bug, we cannot use 1.1.x releases newer than 1.1.5 (or 1.1.4 actually because on another bug that hasn't been fixed yet). If Microsoft releases any security updates or other bug fixes on in 1.1.x in the future, we will not be able to use that _until_ this bug is resolved. This goes against "LTS" principles

**Comment by damonbarry** [comment]
@robinmholt There's a PR (link is above your comment) which we're considering taking for the next 1.1 release.

**Comment by veyalla** [comment]
Thanks all for your input. Based on your feedback a fix for this issue went out yesterday in the 1.1.8 release.
